pkg/xds/envoy/listeners/v3/dns_configurer.go

```diff
@@ -0,0 +1,99 @@
+package v3
+
+import (
+	"sort"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_data_dns "github.com/envoyproxy/go-control-plane/envoy/data/dns/v3"
+	envoy_dns "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/udp/dns_filter/v3alpha"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	"github.com/kumahq/kuma/pkg/util/proto"
+)
+
+type DNSConfigurer struct {
+	VIPs         map[string]string
+	EmptyDNSPort uint32
+}
+
+func (c *DNSConfigurer) Configure(listener *envoy_listener.Listener) error {
+	pbst, err := proto.MarshalAnyDeterministic(c.dnsFilter())
+	if err != nil {
+		return err
+	}
+
+	listener.ListenerFilters = append(listener.ListenerFilters, &envoy_listener.ListenerFilter{
+		Name: "envoy.filters.udp.dns_filter",
+		ConfigType: &envoy_listener.ListenerFilter_TypedConfig{
+			TypedConfig: pbst,
+		},
+	})
+	return nil
+}
+
+func (c *DNSConfigurer) dnsFilter() *envoy_dns.DnsFilterConfig {
+	var virtualDomains []*envoy_data_dns.DnsTable_DnsVirtualDomain
+	for domain, ip := range c.VIPs {
+		virtualDomains = append(virtualDomains, &envoy_data_dns.DnsTable_DnsVirtualDomain{
+			Name: domain,
+			Endpoint: &envoy_data_dns.DnsTable_DnsEndpoint{
+				EndpointConfig: &envoy_data_dns.DnsTable_DnsEndpoint_AddressList{
+					AddressList: &envoy_data_dns.DnsTable_AddressList{
+						Address: []string{
+							ip,
+						},
+					},
+				},
+			},
+			AnswerTtl: nil,
```

Comment:
Should we care about TTL? What if we scale down some service to 0 instances and then scale up it back. I think VIP will be changed?

Reply:
good point, changed to 1s